docs/configuration.rst
.. _configuration:

Configuration
=============

twtxt uses a simple INI-like configuration file. It’s recommended to use ``twtxt quickstart`` to create it. On Linux twtxt checks ``~/.config/twtxt/config`` for its configuration. OSX uses ``~/Library/Application Support/twtxt/config``. Consult `get_app_dir <http://click.pocoo.org/6/api/#click.get_app_dir>`_ to find out the config directory for other operating systems.

Here’s an example ``conf`` file, showing every currently supported option:

.. code::

    [twtxt]
    nick = buckket
    twtfile = ~/twtxt.txt
    twturl = http://example.org/twtxt.txt
    check_following = True
    use_pager = False
    porcelain = False
    limit_timeline = 20
    timeout = 5.0
    sorting = descending
    post_tweet_hook = "scp {twtfile} buckket@example.org:~/public_html/twtxt.txt"
    # post_tweet_hook = "aws s3 {twtfile} s3://mybucket.org/twtxt.txt --acl public-read --storage-class REDUCED_REDUNDANCY --cache-control 'max-age=60,public'"

    [following]
    bob = https://example.org/bob.txt
    alice = https://example.org/alice.txt


[twtxt] section
---------------

+-------------------+-------+------------+---------------------------------------------------+
| Option:           | Type: | Default:   | Help:                                             |
+===================+=======+============+===================================================+
| nick              | TEXT  |            | your nick, will be displayed in your timeline     |
+-------------------+-------+------------+---------------------------------------------------+
| twtfile           | PATH  |            | path to your local twtxt file                     |
+-------------------+-------+------------+---------------------------------------------------+
| twturl            | TEXT  |            | URL to your public twtxt file                     |
+-------------------+-------+------------+---------------------------------------------------+
| check_following   | BOOL  | True       | try to resolve URLs when listing followings       |
+-------------------+-------+------------+---------------------------------------------------+
| use_pager         | BOOL  | False      | use a pager (less) to display your timeline       |
+-------------------+-------+------------+---------------------------------------------------+
| porcelain         | BOOL  | False      | style output in an easy-to-parse format           |
+-------------------+-------+------------+---------------------------------------------------+
| limit_timeline    | INT   | 20         | limit amount of tweets shown in your timeline     |
+-------------------+-------+------------+---------------------------------------------------+
| timeout           | FLOAT | 5.0        | maximal time a http request is allowed to take    |
+-------------------+-------+------------+---------------------------------------------------+
| sorting           | TEXT  | descending | sort timeline either descending or ascending      |
+-------------------+-------+------------+---------------------------------------------------+
| post_tweet_hook   | TEXT  |            | command to be executed after tweeting             |
+-------------------+-------+------------+---------------------------------------------------+

``post_tweet_hook`` is very useful if you want to push your twtxt file to a remote (web) server. Check the example above tho see how it’s used with ``scp``.

[followings] section
--------------------

This section holds all your followings as nick, URL pairs. You can edit this section manually or use the ``follow``/``unfollow`` commands of twtxt for greater comfort.


docs/user/configuration.rst
.. _configuration:

Configuration
=============

twtxt uses a simple INI-like configuration file. It’s recommended to use ``twtxt quickstart`` to create it. On Linux twtxt checks ``~/.config/twtxt/config`` for its configuration. OSX uses ``~/Library/Application Support/twtxt/config``. Consult :func:`click.get_app_dir` to find out the config directory for other operating systems.

Here’s an example ``conf`` file, showing every currently supported option:

.. code-block:: ini

    [twtxt]
    nick = buckket
    twtfile = ~/twtxt.txt
    twturl = http://example.org/twtxt.txt
    check_following = True
    use_pager = False
    use_cache = True
    porcelain = False
    disclose_identity = False
    character_limit = 140
    limit_timeline = 20
    timeout = 5.0
    sorting = descending
    pre_tweet_hook = "scp buckket@example.org:~/public_html/twtxt.txt {twtfile}"
    post_tweet_hook = "scp {twtfile} buckket@example.org:~/public_html/twtxt.txt"
    # post_tweet_hook = "tail -1 {twtfile} | cut -f2 | sed -e 's/^/twt=/'| curl -s -d @- -d 'name=foo' -d 'password=bar' http://htwtxt.plomlompom.com/feeds"
    # post_tweet_hook = "aws s3 cp {twtfile} s3://mybucket.org/twtxt.txt --acl public-read --storage-class REDUCED_REDUNDANCY --cache-control 'max-age=60,public'"

    [following]
    bob = https://example.org/bob.txt
    alice = https://example.org/alice.txt

[twtxt]
-------

+-------------------+-------+------------+---------------------------------------------------+
| Option:           | Type: | Default:   | Help:                                             |
+===================+=======+============+===================================================+
| nick              | TEXT  |            | your nick, will be displayed in your timeline     |
+-------------------+-------+------------+---------------------------------------------------+
| twtfile           | PATH  |            | path to your local twtxt file                     |
+-------------------+-------+------------+---------------------------------------------------+
| twturl            | TEXT  |            | URL to your public twtxt file                     |
+-------------------+-------+------------+---------------------------------------------------+
| check_following   | BOOL  | True       | try to resolve URLs when listing followings       |
+-------------------+-------+------------+---------------------------------------------------+
| use_pager         | BOOL  | False      | use a pager (less) to display your timeline       |
+-------------------+-------+------------+---------------------------------------------------+
| use_cache         | BOOL  | True       | cache remote twtxt files locally                  |
+-------------------+-------+------------+---------------------------------------------------+
| porcelain         | BOOL  | False      | style output in an easy-to-parse format           |
+-------------------+-------+------------+---------------------------------------------------+
| disclose_identity | BOOL  | False      | include nick and twturl in twtxt’s user-agent     |
+-------------------+-------+------------+---------------------------------------------------+
| character_limit   | INT   | None       | limit amount of characters a tweet can have       |
+-------------------+-------+------------+---------------------------------------------------+
| limit_timeline    | INT   | 20         | limit amount of tweets shown in your timeline     |
+-------------------+-------+------------+---------------------------------------------------+
| timeout           | FLOAT | 5.0        | maximal time a http request is allowed to take    |
+-------------------+-------+------------+---------------------------------------------------+
| sorting           | TEXT  | descending | sort timeline either descending or ascending      |
+-------------------+-------+------------+---------------------------------------------------+
| pre_tweet_hook    | TEXT  |            | command to be executed before tweeting            |
+-------------------+-------+------------+---------------------------------------------------+
| post_tweet_hook   | TEXT  |            | command to be executed after tweeting             |
+-------------------+-------+------------+---------------------------------------------------+

``pre_tweet_hook`` and ``post_tweet_hook`` are very useful if you want to push your twtxt file to a remote (web) server. Check the example above tho see how it’s used with ``scp``.

[followings]
------------
This section holds all your followings as nick, URL pairs. You can edit this section manually or use the ``follow``/``unfollow`` commands of twtxt for greater comfort.

CHANGES
Changelog
=========

These are all the changes in twtxt since the first public release.

1.2.1
-----

- Fixed a bug in the mention handling code
- Nicks are now case-insensitive throughout twtxt
- Added a character_limit option, to specify when to shorten tweets
- Added option to include user’s nick and url in twtxt’s user-agent string for greater discovery

1.2.0
-----

- twtxt package is now executable
- twtxt is now available as wheel archive
- Check status of feed before following a new source
- Added a source option to timeline command to only show one specific source
- Added a view command which is basically an alias for the source option
- Added support for local caching of remote twtxt files
- Added support for mentions, see README.rst for more information
- Added a pre_tweet_hook option, works like post_tweet_hook
- Added a porcelain option, styles output script-friendly

1.1.0
-----

- Ignore status updates coming form the future.
- Improved timestamp format by by removing microseconds and using local time.
- Get confirmation before overwriting an existing source.
- Allow tweet command to read from stdin. (echo "Test" | twtxt tweet)
- Allow tweet command to accept unlimited amount of arguments. (twtxt tweet This is  also valid)
- Added an option to change the timeline sorting.
- Added an option to specify max wait time for HTTP requests.


1.0.1
-----

- use subprocess.call() instead of subprocess.run() which is Python 3.5 only.

1.0.0
-----

Initial public release.

twtxt/config.py
"""
    twtxt.config
    ~~~~~~~~~~~~

    This module implements the config file parser/writer.

    :copyright: (c) 2016 by buckket.
    :license: MIT, see LICENSE for more details.
"""

import configparser
import logging
import os

import click

from twtxt.models import Source

logger = logging.getLogger(__name__)


class Config:
    """:class:`Config` interacts with the configuration file.

    :param str config_file: full path to the loaded config file
    :param ~configparser.ConfigParser cfg: a :class:`~configparser.ConfigParser` object with config loaded
    """
    config_dir = click.get_app_dir("twtxt")
    config_name = "config"

    def __init__(self, config_file, cfg):
        self.config_file = config_file
        self.cfg = cfg

    @classmethod
    def from_file(cls, file):
        """Try loading given config file.

        :param str file: full path to the config file to load
        """
        if not os.path.exists(file):
            raise ValueError("Config file not found.")

        cfg = configparser.ConfigParser()

        try:
            cfg.read(file)
            return cls(file, cfg)
        except configparser.Error:
            raise ValueError("Config file is invalid.")

    @classmethod
    def discover(cls):
        """Make a guess about the config file location an try loading it."""
        file = os.path.join(Config.config_dir, Config.config_name)
        return cls.from_file(file)

    @classmethod
    def create_config(cls, nick, twtfile, add_news):
        """Create a new config file at the default location.

        :param str nick: nickname to use for own tweets
        :param str twtfile: path to the local twtxt file
        :param bool add_news: if true follow twtxt news feed
        """
        if not os.path.exists(Config.config_dir):
            os.makedirs(Config.config_dir)
        file = os.path.join(Config.config_dir, Config.config_name)

        cfg = configparser.ConfigParser()

        cfg.add_section("twtxt")
        cfg.set("twtxt", "nick", nick)
        cfg.set("twtxt", "twtfile", twtfile)
        cfg.set("twtxt", "character_limit", "140")

        cfg.add_section("following")
        if add_news:
            cfg.set("following", "twtxt", "https://buckket.org/twtxt_news.txt")

        conf = cls(file, cfg)
        conf.write_config()
        return conf

    def write_config(self):
        """Writes `self.cfg` to `self.config_file`."""
        with open(self.config_file, "w") as config_file:
            self.cfg.write(config_file)

    @property
    def following(self):
        """A :class:`list` of all :class:`Source` objects."""
        following = []
        try:
            for (nick, url) in self.cfg.items("following"):
                source = Source(nick, url)
                following.append(source)
        except configparser.NoSectionError as e:
            logger.debug(e)

        return following

    @property
    def options(self):
        """A :class:`dict` of all config options."""
        try:
            return dict(self.cfg.items("twtxt"))
        except configparser.NoSectionError as e:
            logger.debug(e)
            return {}

    @property
    def nick(self):
        return self.cfg.get("twtxt", "nick", fallback=os.environ.get("USER", "").lower())

    @property
    def twtfile(self):
        return os.path.expanduser(self.cfg.get("twtxt", "twtfile", fallback="twtxt.txt"))

    @property
    def twturl(self):
        return self.cfg.get("twtxt", "twturl", fallback=None)

    @property
    def check_following(self):
        return self.cfg.getboolean("twtxt", "check_following", fallback=True)

    @property
    def use_pager(self):
        return self.cfg.getboolean("twtxt", "use_pager", fallback=False)

    @property
    def use_cache(self):
        return self.cfg.getboolean("twtxt", "use_cache", fallback=True)

    @property
    def porcelain(self):
        return self.cfg.getboolean("twtxt", "porcelain", fallback=False)

    @property
    def disclose_identity(self):
        return self.cfg.getboolean("twtxt", "disclose_identity", fallback=False)

    @property
    def character_limit(self):
        return self.cfg.getint("twtxt", "character_limit", fallback=None)

    @property
    def limit_timeline(self):
        return self.cfg.getint("twtxt", "limit_timeline", fallback=20)

    @property
    def timeout(self):
        return self.cfg.getfloat("twtxt", "timeout", fallback=5.0)

    @property
    def sorting(self):
        return self.cfg.get("twtxt", "sorting", fallback="descending")

    @property
    def source(self):
        return Source(self.nick, self.twturl)

    @property
    def pre_tweet_hook(self):
        return self.cfg.get("twtxt", "pre_tweet_hook", fallback=None)

    @property
    def post_tweet_hook(self):
        return self.cfg.get("twtxt", "post_tweet_hook", fallback=None)

    def add_source(self, source):
        """Adds a new :class:`Source` to the config’s following section."""
        if not self.cfg.has_section("following"):
            self.cfg.add_section("following")

        self.cfg.set("following", source.nick, source.url)
        self.write_config()

    def get_source_by_nick(self, nick):
        """Returns the :class:`Source` of the given nick.

        :param str nick: nickname for which will be searched in the config
        """
        url = self.cfg.get("following", nick, fallback=None)
        return Source(nick, url) if url else None

    def remove_source_by_nick(self, nick):
        """Removes a :class:`Source` form the config’s following section.

        :param str nick: nickname for which will be searched in the config
        """
        if not self.cfg.has_section("following"):
            return False

        ret_val = self.cfg.remove_option("following", nick)
        self.write_config()
        return ret_val

    def build_default_map(self):
        """Maps config options to the default values used by click, returns :class:`dict`."""
        default_map = {
            "following": {
                "check": self.check_following,
                "timeout": self.timeout,
                "porcelain": self.porcelain,
            },
            "tweet": {
                "twtfile": self.twtfile,
            },
            "timeline": {
                "pager": self.use_pager,
                "cache": self.use_cache,
                "limit": self.limit_timeline,
                "timeout": self.timeout,
                "sorting": self.sorting,
                "porcelain": self.porcelain,
                "twtfile": self.twtfile,
            },
            "view": {
                "pager": self.use_pager,
                "cache": self.use_cache,
                "limit": self.limit_timeline,
                "timeout": self.timeout,
                "sorting": self.sorting,
                "porcelain": self.porcelain,
            }
        }
        return default_map
